proc/merge.go

```diff
@@ -16,47 +14,55 @@ import (
 // and this matches that.
 type Merge struct {
 	Base
-	once    sync.Once
-	parents []*runnerProc
-	bufs    []zbuf.Batch
-	err     error
+	once     sync.Once
+	ch       <-chan Result
+	parents  []*runnerProc
+	nparents int
 }
 
 type runnerProc struct {
 	Base
-	ch      chan Result
-	proceed chan struct{}
+	ch     chan<- Result
+	doneCh chan struct{}
 }
 
-func newrunnerProc(c *Context, parent Proc) *runnerProc {
+func newrunnerProc(c *Context, parent Proc, ch chan<- Result) *runnerProc {
 	return &runnerProc{
-		Base:    Base{Context: c, Parent: parent},
-		ch:      make(chan Result),
-		proceed: make(chan struct{}),
+		Base:   Base{Context: c, Parent: parent},
+		ch:     ch,
+		doneCh: make(chan struct{}, 1),
```

Comment:
There's no need for a separate done channel per runnerProc. Just add `doneCh chan struct{}` to Merge, pass it to newrunnerProc where you then select on <-doneCh, and do `close(m.doneCh)` in (*Merge).Done.

Reply:
That's much nicer, thanks